app/domain/README.md

```diff
@@ -0,0 +1,296 @@
+
+Currently, `app/domain` contains mostly authentication logic.  But its scope is
+not limited to authenticators.
+
+At the highest level, this approach aims for:
+
+- Explicit, maintainable code that is easy to understand
+- Fast unit tests
+- Intuitive names that correspond with our documentation and the way we
+  naturally talk
+
+We'll dig into the technical specifics below.
+
+## The "Functional Object-Oriented" Style
+
+I'll offer a TLDR version here.  For more in depth discussion of the same
+principles, the video [Functional Architecture for the Practical
+Rubyist](https://www.youtube.com/watch?v=7qnsRejCyEQ&feature=youtu.be) is a
+good place to start.
+
+The basic idea is to reap many benefits of functional programming while 
+remaining true to ruby's objected-oriented paradigm.
+
+In brief, this means building features out of three basic kinds of objects, and
+maintaing a strict sepration of responsibilities:
+
+- **Command objects**
+  - These go by many names: service objects, interactors, use cases
+  - Coordinates high-level logic involving multiple collaborators
+    - For example: `CreateUser` might need to coordinate user input validation,
+      a database transaction, a log entry, and a confirmation email
+  - Makes all collaborators explicit as dependencies
+  - Easy to unit test with doubles or mocks
+- **Value objects**
+  - These are essentially data types.
+  - Methods return data the object was initialized with, or transformations of
+    it
+  - For example: `User`, `Address`, `Url`, etc.
+  - Immutable
+  - No side effects
+  - Class names are nouns
+  - Method names are nouns describing what they return
+- **Objects with side effects**
+  - Repositories for performing database transactions
+  - Objects that communicate with network services (`Aws`, `Github`, etc)
+  - Objects that interact with the file system
+  - Class names are usually nouns
+  - Method names are verbs (to emphasize their side-effects)
+    - For example: `github.fetch_issues`
+    - By contrast, a value object would use a name like `github.issues`.  The
+      "fetch" indicates the network side effect and possibility of error.
+
+Other principles include:
+
+- Avoid inheritance
+  - Instead, make shared code into its own object, which is passed into other
+    objects that need it.  Composition over inheritance.
+- All objects are immutable
+- Careful, consistent naming practices that align with documentation
+
+### Specific Technologies
+
+Writing in the functional OO style does not require any specific gems, though
+many gems make it easier to write value objects and command objects.
+
+In general, there is a tradeoff when using gems:
+
+- Positive: 
+  - They cut down on boilerplate and make your code more concise
+- Negative: 
+  - Every new developer has another extra thing to learn
+  - The code is often not as obvious as plain ruby
+
+You'll encounter two gems when working in Conjur:
+
+- `command_class` for writing command objects
```

Comment:
do you want to move this repo to a cyberark one?

Reply:
Well, all my work on it was done on my own time.  If there's a pressing need to I suppose I would, but it's really no different from our perspective than using any 3rd party gem and we depend on hundreds of them.